# Cleaning and Exploring Data

***

Now that we've grown a bit accustomed to basic operations in python we're going to jump ahead to how we can load in data using Python. When we start dealing with packages you'll notice (especially with <code>pandas</code>) that *programming begins to revolve around the features provided by the package* rather than the basic python introduced to you in the first section. 

In [1]:
import pandas as pd
import numpy as np

In [2]:
#Data to use
demographics_data = '../data/demographics.csv'
cognitive_data = '../data/cognitive_scoring.csv'
neuro_data = '../data/brain_measurements.csv'
reaction_data = '../data/rt_trial_data.csv'

## Basic Pandas Data Exploration

First we'll start off with the simplest of tasks, suppose that I'd like to load in a CSV and get an idea of what the data might look like. Using <code>pandas</code> this is as easy as a one-liner. We can simply load in the data as a table and explore it effortlessly

In [3]:
demo_df = pd.read_csv(demographics_data)

Pandas by default uses a <code>,</code> delimiter, although I can specify it to be anything I want it to be. For example a tab is encoded as <code>\t</code>, a space is encoded as <code>' '</code>, and so on.


demo_df is a **DataFrame** which is a special pandas data-type which contains lots of useful functions. From now on when we say <code>DataFrame.somethinghere</code>, what we actually mean is that you type out (depending on which **DataFrame** we're talking about):

```python
demo_df.somethinghere
```

That means that <code>somethinghere</code> is common to *all **DataFrames**, not just* <code>demo_df</code>

We can view the first $n$ rows of our data using the <code>head</code> property of the data that we loaded in using pandas

In [4]:
demo_df.head()

subject_id Calculated socioeconomic status score  Education score  \
0  h1NuzaHTaJIM6QZi                     10.46819275002584         4.414423   
1  er4E9WZnXGiJCk8r                    10.894777891322148         4.529836   
2  LEm4YtC45KYNwbFg                     9.806142643511784         3.455996   
3  B9sIFZJs6UD51QAm                     9.310570231612841         2.813627   
4  EkWUnI8h1sgopN3a                     8.986733675059984         3.485714   

  Is person case or control? Is person male or female?  
0                    control                    female  
1                    control                    female  
2                       case                    female  
3                    control                      male  
4                    control                    female

Furthermore we could get a quick summary of what all the columns are as well:

In [5]:
demo_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 5 columns):
subject_id                               400 non-null object
Calculated socioeconomic status score    400 non-null object
Education score                          400 non-null float64
Is person case or control?               400 non-null object
Is person male or female?                400 non-null object
dtypes: float64(1), object(4)
memory usage: 15.7+ KB


Finally we can summarize our entire data-table 

<code>DataFrame.info</code> gives us a bunch of columns that we can work with in our dataset. For example suppose I just wanted the age data from our table. I can select the 'Age' column in one of two ways:

1. demo_df.subject_id
2. demo_df['subject_id'] 

***
*Note that this does not work when the columns have spaces! We'll deal with this in a bit*
***

The second method has the advantage in that you can input a *list* of columns to pull out. For example:
```python
columns = ['subject_id','Is person male or female?','Education score']
demo_df[columns] 
```
or

```python
demo_df[ ['subject_id','Is person male or female?','Education score'] ]
```

Both of which will produce equivalent results


In [6]:
#Try selecting multiple columns!
demo_df[ ['subject_id','Is person male or female?','Education score'] ].head()

subject_id Is person male or female?  Education score
0  h1NuzaHTaJIM6QZi                    female         4.414423
1  er4E9WZnXGiJCk8r                    female         4.529836
2  LEm4YtC45KYNwbFg                    female         3.455996
3  B9sIFZJs6UD51QAm                      male         2.813627
4  EkWUnI8h1sgopN3a                    female         3.485714

In [7]:
#We use include='all' since by default when there is a mix of numeric columns (respondent ID) and non-numeric columns
# such as categories, pandas only analyzes the numeric columns.
demo_df.describe(include='all')

subject_id Calculated socioeconomic status score  \
count                400                                   400   
unique               400                                   396   
top     e2w1FkACJju8WMf7                               missing   
freq                   1                                     5   
mean                 NaN                                   NaN   
std                  NaN                                   NaN   
min                  NaN                                   NaN   
25%                  NaN                                   NaN   
50%                  NaN                                   NaN   
75%                  NaN                                   NaN   
max                  NaN                                   NaN   

        Education score Is person case or control? Is person male or female?  
count        400.000000                        400                       400  
unique              NaN                          2                         2  
top                 NaN                       case                    female  
freq                NaN                        200                       203  
mean           3.932924                        NaN                       NaN  
std            1.313882                        NaN                       NaN  
min            0.536629                        NaN                       NaN  
25%            2.951655                        NaN                       NaN  
50%            4.048368                        NaN                       NaN  
75%            4.885143                        NaN                       NaN  
max            7.402765                        NaN                       NaN

## Conditional data selection

What if we wanted to select rows in which the individual's sex is female? This is called conditional selection and is incredibly useful for exploring, cleaning and analyzing data! Conditional selection is typically done through a two step process:

- Step 1: Pick a condition to mask with
- Step 2: Pull out rows/columns that meet this condition

We can do this in the following way:

In [8]:
#Step 1 -- make a conditional mask
female_mask = demo_df['Is person male or female?'] == 'female'
female_mask.head()

0     True
1     True
2     True
3    False
4     True
Name: Is person male or female?, dtype: bool

<code>female_mask</code> is a *conditional series*, meaning that for each **index** (the left-hand number) there is an associated value (True or False). To select rows from the original DataFrame <code>demo_df</code> in which <code>female_mask</code> is <code>True</code>, we use the <code>df.loc</code> function, which works as follows:

In [9]:
female_df = demo_df.loc[female_mask]
female_df.head()

subject_id Calculated socioeconomic status score  Education score  \
0  h1NuzaHTaJIM6QZi                     10.46819275002584         4.414423   
1  er4E9WZnXGiJCk8r                    10.894777891322148         4.529836   
2  LEm4YtC45KYNwbFg                     9.806142643511784         3.455996   
4  EkWUnI8h1sgopN3a                     8.986733675059984         3.485714   
5  jaLH0mWDEjbncGbL                      9.56720981988074         4.564721   

  Is person case or control? Is person male or female?  
0                    control                    female  
1                    control                    female  
2                       case                    female  
4                    control                    female  
5                    control                    female

<code>df.loc</code> also allows use to set columns, the general syntax is:

```python
df.loc[rows,columns]
```
So if we wanted to get the subject ids in which the individual is a female, we can do the following:

In [10]:
demo_df.loc[female_mask,'subject_id'].head()

0    h1NuzaHTaJIM6QZi
1    er4E9WZnXGiJCk8r
2    LEm4YtC45KYNwbFg
4    EkWUnI8h1sgopN3a
5    jaLH0mWDEjbncGbL
Name: subject_id, dtype: object

Notice how every entry in "Is person male or female?" is female. Using a mask we've selected only the rows in which the indivdual is a female. We can confirm that we have selected a subset of the dataframe by looking at the **index** (left-hand side), in which some numbers are skipped meaning that some rows (males) have been skipped

***

NOTE: An alternative way to apply a mask instead of <code>demo_df.loc</code> is as follows:


```python
demo_df[female_mask]
```

The issue with this is that it isn't *explicit whether we are subsetting our data by rows or columns*. For example if we had a column called <code>True</code>, then it might give us weird behaviours! Using <code>df.loc</code> ensures that we *explicitly tell pandas which rows and columns we're choosing*

See https://stackoverflow.com/questions/38886080/python-pandas-series-why-use-loc for more info

***


## Merging in Multiple Data Tables

Sometimes the data that we're working with a spread out across multiple tables. Ideally we'd like to have all this data available in one table for us to work with. Let's load in all the data available to us and take a look at what we're dealing with:

In [11]:
neuro_df = pd.read_csv(neuro_data)
cog_df = pd.read_csv(cognitive_data)

In [12]:
neuro_df.head()

subject_id  Measured brain volume  Measured brain surface area  \
0  55KMlrjdtghHZLfG            1216.321156                  1635.504837   
1  Wcv979tbmAoLUGmB            1294.128556                  1661.312955   
2  cOg2RWhtUekda5BM            1121.656731                  1695.982823   
3  A7GddC92M0N5g3vJ            1340.437608                  1924.084233   
4  XSEVEJq6OsbS2Xsi            1433.331198                  1807.672001   

   Measured brain grey matter thickness  
0                              2.813704  
1                              6.254939  
2                              4.359744  
3                              3.663411  
4                              6.513289

In [13]:
cog_df.head()

subject_id Neurocognitive battery assessment RX201
0  OejOBT1PoKoG19GJ                        84.0463015666604
1  5DKHOuVYhlDhnaY1                       73.98598404203868
2  HHq8g0sZVCfWVdcg                        84.5713573062133
3  uqBMSpUMV7YxcRNA                       77.18019498058791
4  M2XZEKPhL4g6gXv2                       56.00006052447458

Looks like these dataframes store different types of data! Our final exploration of the data will be reliant on being able to integrate across these differnet measurements. We can perform merging of this data through a process called *merging* or *joining* which is accomplished through <code>pd.merge</code>. Table merging relies on the ability to link separate data tables through a common **key**. 

The **key** is described as a column of data which uniquely identifies each row. In our case <code>subject_id</code> is a unique key across each of our data tables. Therefore we'd like to *merge our data on the <code>subject_id</code> key*. There are two ways to accomplish this:

1. <code>pd.merge(df1,df2,...)</code> 
2. <code>df1.merge(df2,...)</code>


Both methods are equivalent and method 2 saves a bit of typing if you're the lazy kind.

In [14]:
#Merge all dataframes
df = demo_df.merge(neuro_df,on='subject_id').merge(cog_df,on='subject_id')
df.head()

subject_id Calculated socioeconomic status score  Education score  \
0  h1NuzaHTaJIM6QZi                     10.46819275002584         4.414423   
1  er4E9WZnXGiJCk8r                    10.894777891322148         4.529836   
2  LEm4YtC45KYNwbFg                     9.806142643511784         3.455996   
3  B9sIFZJs6UD51QAm                     9.310570231612841         2.813627   
4  EkWUnI8h1sgopN3a                     8.986733675059984         3.485714   

  Is person case or control? Is person male or female?  Measured brain volume  \
0                    control                    female            1416.994066   
1                    control                    female            1321.507702   
2                       case                    female            1056.932445   
3                    control                      male            1303.358436   
4                    control                    female             951.380073   

   Measured brain surface area  Measured brain grey matter thickness  \
0                  1919.613895                              8.715702   
1                  1895.997272                              6.516022   
2                  1690.701079                              3.569380   
3                  1961.765667                              4.947032   
4                  1683.487804                              4.967411   

  Neurocognitive battery assessment RX201  
0                       82.44403200496558  
1                       72.24443919868312  
2                       59.43805755950151  
3                       91.61414162964638  
4                       80.19413970263216

***

## Data Pre-processing

Sometimes the data that we acquire isn't in a format that makes sense to directly merge into our DataFrame. The data may be messy, or in a format that needs to be pre-processed in some way in order to be useful to us. Take for example the reaction time dataset:

In [15]:
rt_df = pd.read_csv(reaction_data)
rt_df.head()

p7U1juxknprs2w5C  nd2Y3hGWKQVSHj75  yiMrzXHoelyY7EJM  7Tm46aoyg0a7aSBV  \
0        297.388996        624.544358        193.333440       1639.113721   
1       3080.506797       1082.419826        457.412916        400.587468   
2        757.828822        171.189947       1249.604311        453.522795   
3        932.509308        321.421064        766.086451        105.153265   
4        753.429971        151.801627        905.437426        130.364513   

   uoI3YX4pk5jx398v  5kOElvYqudKlE8Bp  xIUbvr9raKhRBchQ  6wrZ6ukoBfWQg5OL  \
0        358.863112        296.576470        219.897575        290.062739   
1        239.549580        878.193209       1347.321810        616.614208   
2        258.581993       1594.212406       1703.847500        352.556145   
3        360.048778        323.752670        673.058323        203.096848   
4        683.543020       1103.167186       1143.336699        142.067043   

   WJujL13GB1K9QcEC  wPUQ3vuELVkpEQcC        ...         qTvCVQ1vGCtuJHRM  \
0        527.461708        202.253649        ...               795.961253   
1       1827.628849        722.787143        ...              1647.468287   
2        829.798312        876.089111        ...              1324.450413   
3        399.474765        411.244841        ...               728.552762   
4        452.658029       1296.602332        ...              1874.615046   

   zNiporfyXhj9D3gS  WBEjHVNo2VfLmzWM  rDeQWbepzEKqdcVs  jCODs4zdn9PFKAzt  \
0        532.345279        670.139030         39.565514        956.630815   
1       1100.057318        945.467636       1399.424567       2528.348480   
2        758.214912        318.801871       1747.286436        811.551506   
3       1217.433509        247.809811        176.611345        368.144410   
4       1236.868205       2226.146076       1126.258708        275.221173   

   cnrdgS42LprKpOEi  oNwDHbCbgtZRZQ4X  It90LdEUi6E6RuDR  kahbyz0El6rukGrb  \
0        374.187821        167.447220       2396.595483       1204.517042   
1        268.824695        592.891487       1349.853774       1748.239480   
2        656.965341        410.635970        595.677105        152.551560   
3        326.872324        230.662930       2437.635403        746.850921   
4       2934.747615        967.621029       2308.719778       2369.434237   

   rTxRLB7ELoW24QgC  
0        404.235658  
1        722.158525  
2        482.634082  
3       1230.849790  
4       1424.368797  

[5 rows x 400 columns]

In [16]:
rt_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 250 entries, 0 to 249
Columns: 400 entries, p7U1juxknprs2w5C to rTxRLB7ELoW24QgC
dtypes: float64(400)
memory usage: 781.3 KB


First of all notice the following components:
1. Each column maps onto a subject on our main DataFrame
2. Each row is a recording from a trial, there are 250 entries

The raw reaction times from each trial is noisy and not entirely useful for us. We'd like to take the average of each column. That is, we'd like to *apply* a function which takes the average of each column! Doing this is straightforward with pandas using the <code>DataFrame.apply</code> function! This function generally looks like this:


```python
rt_df.apply(func)
```

What this does is apply any function <code>func</code> onto each *column of the dataframe rt_df*. If we wanted to apply a function to each *row instead* we'd write it as follows:

```python
rt_df.apply(func,axis=1)
```

This will apply a function to each column instead of row! Let's see how this works in practice:

In [17]:
mean_rt = rt_df.apply(np.mean,axis=0)
mean_rt.head()

p7U1juxknprs2w5C    832.271877
nd2Y3hGWKQVSHj75    659.682523
yiMrzXHoelyY7EJM    706.891307
7Tm46aoyg0a7aSBV    726.971312
uoI3YX4pk5jx398v    670.550973
dtype: float64

Here each subject ID has their mean reaction time computed! An alternative way to do this (for means specifically, this won't work for other custom functions) is to use <code>DataFrame.mean()</code>:

In [18]:
other_method = rt_df.mean()
other_method.head()

p7U1juxknprs2w5C    832.271877
nd2Y3hGWKQVSHj75    659.682523
yiMrzXHoelyY7EJM    706.891307
7Tm46aoyg0a7aSBV    726.971312
uoI3YX4pk5jx398v    670.550973
dtype: float64

This returns an equivalent result. Under the hood it's doing the same thing as using <code>DataFrame.apply</code>. Now we need to merge in our data, one might think to use <code>DataFrame.merge</code>, but there is a caveat!


mean_rt is no longer a DataFrame! In fact, it is called a **Series**, which is a list of (index,value) pairings. Remember <code>female_mask</code>? *That was also a **Series** object!*. 

The left-hand side is the index, the right hand-side is the value. So when merging this series into our main DataFrame we need to take this subtlety into consideration! We can get around this by first converting our **Series** into a DataFrame then merging it in as usual

In [19]:
#Convert series to dataframe
mean_rt_df = pd.DataFrame(mean_rt)
mean_rt_df.head()

0
p7U1juxknprs2w5C  832.271877
nd2Y3hGWKQVSHj75  659.682523
yiMrzXHoelyY7EJM  706.891307
7Tm46aoyg0a7aSBV  726.971312
uoI3YX4pk5jx398v  670.550973

Notice how the **index** on the left-hand side is now the subject keys? When we merge DataFrames we want to make sure we specify this since we don't have a column called 'subject_id' as we did before

In [20]:
df = df.merge(mean_rt_df,left_on='subject_id',right_on=mean_rt_df.index)
df.head()

subject_id Calculated socioeconomic status score  Education score  \
0  h1NuzaHTaJIM6QZi                     10.46819275002584         4.414423   
1  er4E9WZnXGiJCk8r                    10.894777891322148         4.529836   
2  LEm4YtC45KYNwbFg                     9.806142643511784         3.455996   
3  B9sIFZJs6UD51QAm                     9.310570231612841         2.813627   
4  EkWUnI8h1sgopN3a                     8.986733675059984         3.485714   

  Is person case or control? Is person male or female?  Measured brain volume  \
0                    control                    female            1416.994066   
1                    control                    female            1321.507702   
2                       case                    female            1056.932445   
3                    control                      male            1303.358436   
4                    control                    female             951.380073   

   Measured brain surface area  Measured brain grey matter thickness  \
0                  1919.613895                              8.715702   
1                  1895.997272                              6.516022   
2                  1690.701079                              3.569380   
3                  1961.765667                              4.947032   
4                  1683.487804                              4.967411   

  Neurocognitive battery assessment RX201           0  
0                       82.44403200496558  870.207541  
1                       72.24443919868312  713.317221  
2                       59.43805755950151  970.323949  
3                       91.61414162964638  763.525358  
4                       80.19413970263216  755.453697

That's it! We'll deal with the ugly $0$ column name in a bit :)

## Data Cleaning

Data cleaning is a multi-step process that typically consists of the following steps:

1. Renaming columns so that they're easier to work with
2. Dealing NaN values or "Impossible values" (for example sex = -20) 
4. Removing columns that you're not interested in analyzing

The process of data cleaning may be extensive and in many cases *is the most extensive part* of analyzing data. In this tutorial we'll quickly go over how one might perform each of the above steps using pandas and end off with a clean dataset that we can start exploring and visualizing.

One key point is that data cleaning is sometimes easier if you visualize your data! This is a great thing to do when looking at whether you're dealing with impossible values that should not exist in your dataset

### Renaming Columns

In some cases the data that we acquire doesn't come in a format that make sense to merge into our dataframe. Either the data is too messy, is in a raw format that isn't directly useful for visualization and analysis or 

There are several reasons why one might want to rename their columns in an analysis. Here are a few:


**Some columns just don't make any sense!** Take a look at one of the columns we're dealing with in our dataset called *Unnamed: 30*. Now tell me what exactly this column is supposed to tell you without looking through the dataset. *You can't*, instead you'll have to do the hard-work of deciphering this or writing it down in your notebook so you can keep track of what columns mean. In this case the solution is to first decipher the column, and there are several ways we could do this, then to *rename the column to one that makes much more sense and aligns with what the data indicates*. 


Here's another one: Our dataset has a column called 'How many days were you hospitalized for your mental illness'. You could pull this column from your dataset in the following way:

```python
df['Is person male or female?']
```

But typiing this out repeatedly is a pain and gets tiring quick. Life would be SO MUCH EASIER if we could just rename this column to access our data like this

```python
df['sex']
```

This is succint and gets to the point about what the variable means in the context of this data.

***

Let's tackle the first problem. First let's grab a list of columns:

In [21]:
columns = df.columns

In [22]:
print(columns)

Index([                             'subject_id',
         'Calculated socioeconomic status score',
                               'Education score',
                    'Is person case or control?',
                     'Is person male or female?',
                         'Measured brain volume',
                   'Measured brain surface area',
          'Measured brain grey matter thickness',
       'Neurocognitive battery assessment RX201',
                                               0],
      dtype='object')


In [23]:
shorten_map = {
    'Calculated socioeconomic status score': 'se_score',
    'Education score' : 'edu_score',
    'Is person case or control?' : 'group',
    'Is person male or female?' : 'sex',
    'Measured brain volume' : 'brain_vol',
    'Measured brain surface area': 'brain_surfarea',
    'Measured brain grey matter thickness' : 'brain_gm_thick',
    'Neurocognitive battery assessment RX201' : 'cog_score',
    0 : 'mean_rt'
}

In [24]:
df = df.rename(shorten_map,axis=1)

In [25]:
df.head()

subject_id            se_score  edu_score    group     sex  \
0  h1NuzaHTaJIM6QZi   10.46819275002584   4.414423  control  female   
1  er4E9WZnXGiJCk8r  10.894777891322148   4.529836  control  female   
2  LEm4YtC45KYNwbFg   9.806142643511784   3.455996     case  female   
3  B9sIFZJs6UD51QAm   9.310570231612841   2.813627  control    male   
4  EkWUnI8h1sgopN3a   8.986733675059984   3.485714  control  female   

     brain_vol  brain_surfarea  brain_gm_thick          cog_score     mean_rt  
0  1416.994066     1919.613895        8.715702  82.44403200496558  870.207541  
1  1321.507702     1895.997272        6.516022  72.24443919868312  713.317221  
2  1056.932445     1690.701079        3.569380  59.43805755950151  970.323949  
3  1303.358436     1961.765667        4.947032  91.61414162964638  763.525358  
4   951.380073     1683.487804        4.967411  80.19413970263216  755.453697

### Dealing with Impossible Values

Data entry is a task that is incredibly prone to error. Sometimes data is collected incorrectly resulting in values that should not exist or are useful for helping us understand the phenomena of interest. First let's lay out some rules associated with our data that we *should know a priori*. 

- <code>subject_id</code> should be a random alphanumeric string
- <code>se_score</code> is a value <100
- <code>edu_score</code> is a value <100
- <code>sex</code> male, female, or other
- <code>group</code> control or case
- <code>brain_vol</code> some floating point number that is positive
- <code>brain_surfarea</code> some floating point number that is positive
- <code>brain_gm_thick</code> some floating point number that is positive but not "substantially big"
- <code>cog_score</code> some floating point number that is positive

Let's quickly check over what kinds of values we have!

In [26]:
df.head()

subject_id            se_score  edu_score    group     sex  \
0  h1NuzaHTaJIM6QZi   10.46819275002584   4.414423  control  female   
1  er4E9WZnXGiJCk8r  10.894777891322148   4.529836  control  female   
2  LEm4YtC45KYNwbFg   9.806142643511784   3.455996     case  female   
3  B9sIFZJs6UD51QAm   9.310570231612841   2.813627  control    male   
4  EkWUnI8h1sgopN3a   8.986733675059984   3.485714  control  female   

     brain_vol  brain_surfarea  brain_gm_thick          cog_score     mean_rt  
0  1416.994066     1919.613895        8.715702  82.44403200496558  870.207541  
1  1321.507702     1895.997272        6.516022  72.24443919868312  713.317221  
2  1056.932445     1690.701079        3.569380  59.43805755950151  970.323949  
3  1303.358436     1961.765667        4.947032  91.61414162964638  763.525358  
4   951.380073     1683.487804        4.967411  80.19413970263216  755.453697

1. Check numeric types:

- se_score
- edu_score
- brain_vol
- brain_surfarea
- brai_gm_thick
- cog_score

First we can do an overall check on our data. Recall that <code>df.info</code> gives us a high level overview of our data. We can expect that all numeric data (listed above) should be of a numeric data type

In [27]:
numeric_types = ['se_score','edu_score','brain_vol','brain_surfarea','brain_gm_thick','cog_score'] 
df[numeric_types].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 400 entries, 0 to 399
Data columns (total 6 columns):
se_score          400 non-null object
edu_score         400 non-null float64
brain_vol         400 non-null float64
brain_surfarea    400 non-null float64
brain_gm_thick    400 non-null float64
cog_score         400 non-null object
dtypes: float64(4), object(2)
memory usage: 21.9+ KB


Notice how:

- se_score
- cog_score

Are non-numeric. This is the first indication that there might be something up with our data. Let's inspect this a bit closer by examining what data is not numeric. 

First we need to know how to determine whether a value is a number or not. We can do this in many ways, but the method with the least hassle is to use python type-conversion. Recall that we can convert a variable from <code>string</code> to <code>float</code> using the python function <code>float()</code>. Also remember that *python will complain with a <code>ValueError</code>* if it cannot convert the <code>string</code> data into <code>float</code>. Turns out that we can actually use this to our advantage using the nifty <code>try/catch</code> functionality. Let's wrap this in a function:

In [28]:
def is_number_numeric(x):
    
    try:
        float(x)
    except ValueError:
        return False
    else:
        return True

Now we want to *apply* this function <code>is_number_numeric</code>, to each element of our dataframe for our two suspect columns. It turns out that to *apply* a function to each *element* it's as easy as using the function <code>df.applymap</code>. <code>df.applymap(func)</code>, applies the function <code>func</code> to each element of a dataframe <code>df</code> and returns a new <code>DataFrame</code> with each element having the function applied to. This is easier to see in practice, let's use <code>applymap</code> to determine whether each number is numeric!

In [29]:
numeric_numbers = df[['se_score','cog_score']].applymap(is_number_numeric)
numeric_numbers.head()

se_score  cog_score
0      True       True
1      True       True
2      True       True
3      True       True
4      True       True

Now we can use **conditional masking** in order to find rows in which the element cannot be converted into a <code>float</code>! 

In [30]:
se_score_mask = numeric_numbers['se_score'] == False
se_score_mask.head()

0    False
1    False
2    False
3    False
4    False
Name: se_score, dtype: bool

In [31]:
df.loc[se_score_mask]

subject_id se_score  edu_score    group     sex    brain_vol  \
10   N270JLbSp7gSknwa  missing   4.051433  control    male  1433.232040   
67   jEXLO78WvFrV6ZvH  missing   3.130900     case  female  1157.036298   
128  DfxFYAnbVXUNwG4L  missing   2.330065     case  female  1054.202508   
148  2hD1nvZ8yubeAJZQ  missing   4.067032     case  female  1128.513513   
191  c2q3pJr5iqR3Y0nq  missing   2.810406     case  female  1203.930276   

     brain_surfarea  brain_gm_thick           cog_score     mean_rt  
10      1691.251832        8.385789   63.32841427415659  722.223792  
67      1913.360408        3.280578  54.268254119115674  851.100871  
128     1733.823976        2.866103  49.960716656966945  864.267959  
148     1751.795571        3.069786   57.67836256297097  833.317985  
191     1843.758658        2.415522   54.96200471462524  830.564846

Here's a one-liner to achieve the same thing!

In [32]:
df.loc[numeric_numbers['se_score'] == False,'se_score']

10     missing
67     missing
128    missing
148    missing
191    missing
Name: se_score, dtype: object

In [33]:
df.loc[numeric_numbers['cog_score'] == False,'cog_score']

8      refused
18     refused
56     refused
177    refused
180    refused
182    refused
201    refused
218    refused
383    refused
Name: cog_score, dtype: object

Turns out that the values in <code>se_score</code> and <code>cog_score</code> that cannot be converted are 'missing' and 'refused' respectively. These values are effectively useless for our analysis, so we're going to have to replace them with **something**. Now there are two options here when trying to fill in missing or nonsense values:


1. Replace them with a value which signals that we should ignore it and proceed on as usual
2. Estimate what the value could be based on other information available to us

This is the problem of **imputation** which is an active area of research in statistical modelling. We won't dive too deep into this as it is a course all on it's own. For now we'll replace the data with a "nan" value which is a special type of value provided for by numpy. The "nan" value indicates that this value is meaningless and should be ignored. 

To do this we can use masking to select and replace rows!

In [34]:
df.loc[numeric_numbers['se_score'] == False,'se_score'] = np.nan
df.loc[numeric_numbers['cog_score'] == False,'cog_score'] = np.nan

#View our changes as we did before!
df.loc[numeric_numbers['se_score'] == False].head()

subject_id se_score  edu_score    group     sex    brain_vol  \
10   N270JLbSp7gSknwa      NaN   4.051433  control    male  1433.232040   
67   jEXLO78WvFrV6ZvH      NaN   3.130900     case  female  1157.036298   
128  DfxFYAnbVXUNwG4L      NaN   2.330065     case  female  1054.202508   
148  2hD1nvZ8yubeAJZQ      NaN   4.067032     case  female  1128.513513   
191  c2q3pJr5iqR3Y0nq      NaN   2.810406     case  female  1203.930276   

     brain_surfarea  brain_gm_thick           cog_score     mean_rt  
10      1691.251832        8.385789   63.32841427415659  722.223792  
67      1913.360408        3.280578  54.268254119115674  851.100871  
128     1733.823976        2.866103  49.960716656966945  864.267959  
148     1751.795571        3.069786   57.67836256297097  833.317985  
191     1843.758658        2.415522   54.96200471462524  830.564846

An alternative way to fill in these missing values is to use the function <code>pd.to_numeric</code>. This will in effect convert all the data in the specified columns into a numeric type and replace an non-numeric values with a NaN. Why didn't we just use this from the get-go? Because it's important to know *what data you're replacing with NaN first before doing so!* Let's convert se_score and cog_score into numeric types:

In [35]:
df.loc[:, ['se_score','cog_score']] = df.loc[:, ['se_score','cog_score']].apply(pd.to_numeric)

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 400 entries, 0 to 399
Data columns (total 10 columns):
subject_id        400 non-null object
se_score          395 non-null float64
edu_score         400 non-null float64
group             400 non-null object
sex               400 non-null object
brain_vol         400 non-null float64
brain_surfarea    400 non-null float64
brain_gm_thick    400 non-null float64
cog_score         391 non-null float64
mean_rt           400 non-null float64
dtypes: float64(7), object(3)
memory usage: 34.4+ KB


Problem fixed! Now let's fix the rest of the dataset is correct using conditional statements as we've done above!

First we check that se_score, edu_score, and brain_gm_thick are not large. Let's say anything greater than 100

In [37]:
too_big = df[ ['se_score','edu_score','brain_gm_thick'] ] > 100
too_big.head()

se_score  edu_score  brain_gm_thick
0     False      False           False
1     False      False           False
2     False      False           False
3     False      False           False
4     False      False           False

Now let's apply the fix as we've done before

In [38]:
df.loc[too_big['se_score'] == True, 'se_score'] = np.nan
df.loc[too_big['edu_score'] == True, 'edu_score'] = np.nan
df.loc[too_big['brain_gm_thick'] == True,'brain_gm_thick'] = np.nan

In [39]:
df.head()

subject_id   se_score  edu_score    group     sex    brain_vol  \
0  h1NuzaHTaJIM6QZi  10.468193   4.414423  control  female  1416.994066   
1  er4E9WZnXGiJCk8r  10.894778   4.529836  control  female  1321.507702   
2  LEm4YtC45KYNwbFg   9.806143   3.455996     case  female  1056.932445   
3  B9sIFZJs6UD51QAm   9.310570   2.813627  control    male  1303.358436   
4  EkWUnI8h1sgopN3a   8.986734   3.485714  control  female   951.380073   

   brain_surfarea  brain_gm_thick  cog_score     mean_rt  
0     1919.613895        8.715702  82.444032  870.207541  
1     1895.997272        6.516022  72.244439  713.317221  
2     1690.701079        3.569380  59.438058  970.323949  
3     1961.765667        4.947032  91.614142  763.525358  
4     1683.487804        4.967411  80.194140  755.453697

At this point you've probably gotten the idea. *We need to extensively check our data before using it for errors!*. I'll just say (since I made up this data), that there are no more issues to resolve. But in reality you would have to perform checks on *all the data you plan to use!*. In any type of analysis workflow, data cleaning is the longest, most boring, but most important component of the workflow! The decisions you make here will influence the rest of your analysis!

Now our dataset is ready to go! Let's visualize some things to learn interesting aspects of our data! Before we do that we need to save our data, this is an easy one-liner

In [40]:
df.to_pickle('../output/cleaned_data.pkl')

**BUT WAIT**: How do we automate this? Let's say we collect more data down the line and we want to spit out a cleaned up dataset. The answer to this is to download the data as a **Python script**. We won't get into automation in this workshop but it's something that is incredibly easy to do once you've gotten a Jupyter notebook prototype like this one set up!